### PR TITLE
docs: add stability index

### LINKED ISSUE #3076: [RFC] Stability Index
**Status:** Open for comments

## Need

Backstage is heading out of alpha and on to a path to an eventual long-term stable release and version `1.0`. The community and ecosystem is growing at an increasing rate, and we want to provide a solid foundation for everyone building things in, with, and around Backstage.

At this stage of the project we also want to be able to move fast and break things. There are many areas where we want to be able to revise design decisions, often based on valuable feedback from the Backstage community! We also want to be able to ship new features and systems but keep iterating and gathering feedback before they reach a stable API.

Because of the rapid evolution combined with the stabilization efforts, there's uncertainty in what parts of the project are stable and which are not. Regardless of whether you're contributing to the project directly, setting up your own Backstage, or building plugins, you will be using various parts of the project with differences in stability. Many API surfaces of Backstage are reasonably solid and are unlikely to see any big breaking changes, while others are much more uncertain and may even have planned breaking changes that haven't shipped yet.

There is clearly room for more guidance both for contributors and users, as well as commitment from the core team to ensure stability of certain parts of the platform.

### Stability Index

We introduce a Stability Index, similar to [the one used by Node.js](https://nodejs.org/dist/latest-v12.x/docs/api/documentation.html#documentation_stability_index). It is a score assigned to a subset of the project, indicating the level of maturity of the API and the commitment to backwards compatibility. However, because of the current phase of the project, and essentially all API surfaces being experimental, we use a slightly different implementation. Instead of the score indicating a perceived stability, a higher score is instead a commitment to providing a smoother upgrade path for users, through both better documentation and backwards compatibility. Importantly, the Stability Index does not supersede [semver](https://semver.org/), meaning we will still adhere to semver and only do breaking changes in minor releases as long as we are on `0.x`.

My proposal is a 0-3 score as follows, each point building on top of the previous one:

0. Breaking changes are noted in the changelog, and documentation is updated.
1. The changelog entry includes a clearly documented upgrade path, providing guidance for how to migrate previous usage patterns to the new version.
2. Breaking changes always include a deprecation phase where both the old and the new APIs can be used in parallel. This deprecation must have been released for at least two weeks before the deprecated API is removed in a minor version bump.
3. The time limit for the deprecation is 3 months instead of two weeks.

TL;DR:

0. There's a changelog entry.
1. There's a migration guide.
2. 2 weeks of deprecation.
3. 3 months of deprecation.

Note that these are commitments to a minimum, as in we may for example still provide a deprecation path even if an API has a stability score of 1.

I am also suggesting that we do not keep this system in place after version `1.0`, at least not in its current form. The suggestion is based on something that makes sense for the current state of the project, and may even evolve before we reach version `1.0`.

There are some intricacies of employing these rules across the board, as there are various different kinds of breakages that behave in different ways. So there are a couple more points to help guide the process:

- Changes that cause breakages outside the main repo without the user doing a version bump should always have a deprecation path. This includes things like changing our example catalog yaml files, which people rely on directly through GitHub.
- Breaking changes only apply to public API. In the case of TypeScript packages, the public API is any types exported by the package, except explicit exceptions like marking a particular function experimental or private.
- In cases where it is not possible to provide a smooth deprecation path for APIs with a score of 2 or above, the breaking change must be as apparent and clear to the user as possible. It might for example include an error being thrown with clear instructions for how to migrate, or even suggest a new code snippet.
- Changes to templates, such as the `create-app` package, are not considered breaking changes. The stability score refers to the interaction with the template itself.

### API Surfaces

For the index to be useful we need to make a sane split of the project in to smaller slices. I suggest that we start the division with each package being its own surface, where some packages are further divided into different sections. We may divide or merge sections as necessary over time, in order to provide the granularity we need.

We also only assign scores to the packages and plugins in the main Backstage repository, and while other plugins are welcome to adopt the Stability Index, it is nothing we would enforce in any way.

Private packages such as the `example-app` or `storybook` are excluded.

### Implementation

A new documentation page is added with a full map of the project, with a list of all packages and sections. Each package or section includes a short description, the stability index, and possibly a motivation. We could also use this file as a place to note deprecations for removal in the future. It might look something like this:

> # Backstage Stability Index
>
> ## Packages
>
> ### @backstage/catalog-model
>
> Contains the core catalog model, for usage in both the frontend and the Backend.
>
> Stability: 2. The catalog model is evolving, but because of the broad usage we want to ensure some stability.
>
> ### @backstage/config
>
> Provides the logic and interfaces for reading static configuration.
>
> Stability: 2
>
> ### @backstage/core
>
> #### Section: React Components
>
> All of the React components exported from `src/components/` and `src/layout/`
>
> Stability: 1
>
> #### Section: Plugin API
>
> The core plugin API and the way plugins expose functionality to Apps.
>
> Stability: 2. There are planned breaking changes around the way that plugins expose components for apps.
>
> Deprecations:
>
> - `v0.2.1` Deprecated the Foo API, it will be removed in a release after 7th Dec 2020.
>
> #### Section: App API
>
> The APIs used in the app, such as `createApp`.
>
> Stability: 2
>
> ...
>
> ### @backstage/dev-utils
>
> Provides utilities for developing plugins in isolation.
>
> Stability: 0. This package is largely broken and needs updates.
>
> ...
>
> ## Plugins
>
> Plugins are rarely marked as stable as the `@backstage/core` plugin API is under evolution.
>
> ### @backstage/catalog
>
> Stability: 1
>
> ...

## Risks

A commitment to stability will also mean greater friction for both contributors and for the core team to manage the project as a whole. For this reason we would need to carefully assess what scores we assign, as over-promising may lead to significant overhead.

Tied to the above risk is the fact that managing the deprecations is a bunch of work. Realistically this is necessary regardless of whether we introduce an API Stability Index or not, as long-term stability is hugely important for a project like this.

Another risk is that the division of the project is too granular, coarse, and/or confusing. I'm hoping we'll be able to work out a good model though, and we can definitely keep evolving it over time too.

## Alternatives

The scoring system is based around our commitment to smoothing out breaking changes. We could take the approach of scoring based on the perceived stability of the API instead, basically 1 = extremely likely to change and 5 = very unlikely to change. I find that it is easy to not be careful enough when assigning that kind of score though, and that a system with some actual implications of each score is more effective in achieving the intended results.

There are many alternatives to the suggested implementation too. We could for example be much more granular and assign a score for type and method exposed by packages in JSDoc, or move the scores to each package README. I feel that a single file with all packages and sections is likely to be the most valuable at this point. It ensures that we keep the total length and time of reading down, while also providing a single point for a quick overview of the project.

### PR DESCRIPTION
**Pull request by Rugvip**
Closes #3076

Some open questions:

- Should we document a more strict process around updating stabilities? The plan in the core team is to target different parts of the project and then update the stability when we feel we're ready to commit to higher stability. Wondering if it's useful with a bit of transparency or if that's not generally interesting to most people reading this doc.
- How do we handle plugins not owned by the core team, should they be included at all, perhaps opt-in from authors? Right now we've grabbed descriptions from the plugin marketplace and set stability to 0.
- Similar to the first point, should we document a process for plugin authors to update the stability? It'd really just be a PR to this doc.

### CODE REVIEW

File: docs/overview/stability-index.md [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+### [tech-radar](https://github.com/backstage/backstage/tree/master/plugins/tech-radar/)
+
+Visualize the your company's official guidelines of different areas of software
+development.
+
+Stability: `0`
+
+### [techdocs](https://github.com/backstage/backstage/tree/master/plugins/techdocs/)
+
+The frontend component of the TechDocs plugin, used to browser technical
```
Comment (hooloovooo) [inline docs/overview/stability-index.md:445]:
```suggestion
The frontend component of the TechDocs plugin, used to browse technical
```

File: .github/styles/vocab.txt [hunk 2]
```diff
@@ -80,6 +82,7 @@ inlinehilite
 interop
 javascript
 Javascript
+jenkins
```
Comment (Fox32) [inline .github/styles/vocab.txt:85]: Btw, this might lead to users writing jenkins instead of Jenkins. I know why added it, as the headings contain the package name. vale might skip that if you write `jenkins` instead (in code block)

File: docs/overview/stability-index.md [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+documentation of entities.
+
+Stability: `1`
+
+### [techdocs-backend](https://github.com/backstage/backstage/tree/master/plugins/techdocs-backend/)
+
+The backend component of the TechDocs plugin, used to transform and serve
+TechDocs.
+
+Stability: `1`
```
Comment (hooloovooo) [inline docs/overview/stability-index.md:455]: Are we in a state where we can commit to providing clear upgrade paths when doing major changes to this? I feel the Reader is stable enough, but i'm not sure about the backend. @backstage/techdocs-core

File: .github/styles/vocab.txt (hunk 2 above)
Comment (Rugvip) [reply .github/styles/vocab.txt]: Yeah think it looks alright with codeblocks, good idea!

File: docs/overview/stability-index.md (hunk 3 above)
Comment (Rugvip) [reply docs/overview/stability-index.md:455]: Could definitely divide things into sections like some of the other packages too. Feel free to push changes to this branch :grin:

File: docs/overview/stability-index.md [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+planned.
+
+### [`catalog-model`](https://github.com/backstage/backstage/tree/master/packages/catalog-model/)
+
+Contains the core catalog model, and utilities for working with entities. Usable
+both in frontend and Backend.
+
+Stability: `2`. The catalog model is evolving, but because of the broad usage we
+
+want to ensure some stability.
```
Comment (dhenneke) [inline docs/overview/stability-index.md:78]: New paragraph intended?

**Comment by Fox32** [comment]
I think it's fine for now, but:

> Right now we've grabbed descriptions from the plugin marketplace and set stability to 0

Maybe the stability index could be part of the machine readable plugin description and could be used to generate this page. However that wouldn't solve the problem for all the internal stuff.

**Comment by stefanalund** [comment]
I think only APIs that are considered part of the platform should be included in the index. As such, any plugin that is not part of the core feature set (catalog, templates, ...) can be excluded. I think this should be the case even if they are part example app. 

What do you think about calculating the current average stability index so that it can be tracked over time?

Since API stability is the top priority right now I think a blog post about the stability index is in order.

File: docs/overview/stability-index.md [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+
+A plugin that visualizes Jenkins workflows for entities. Jenkins offers a simple
+way to set up a continuous integration and continuous delivery environment.
+
+Stability: `0`
+
+### [`kubernetes`](https://github.com/backstage/backstage/tree/master/plugins/kubernetes/)
+
+The frontend component of the Kubernetes plugin, used to browse and visualize
+Kubernetes resources. Stability: `1`.
```
Comment (adamdmharvey) [inline docs/overview/stability-index.md:352]:
```suggestion
Kubernetes resources.

Stability: `1`.
```

Comment (adamdmharvey) [reply docs/overview/stability-index.md:352]: For formatting consistency.

File: docs/overview/stability-index.md [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+
+### [`kubernetes-backend`](https://github.com/backstage/backstage/tree/master/plugins/kubernetes-backend/)
+
+The backend component of the Kubernetes plugin, used to fetch Kubernetes
+resources from clusters and associate them with entities in the Catalog.
+Stability: `1`.
```
Comment (adamdmharvey) [inline docs/overview/stability-index.md:358]:
```suggestion

Stability: `1`.
```

File: docs/overview/stability-index.md [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+Google's Lighthouse tool is a great resource for benchmarking and improving the
+accessibility, performance, SEO, and best practices of your website.
+
+Stability: `0`
+
+### [`newrelic`](https://github.com/backstage/backstage/tree/master/plugins/newrelic/)
+
+Observability platform built to help engineers create and monitor their
+software. Stability: `0`
```
Comment (adamdmharvey) [inline docs/overview/stability-index.md:370]:
```suggestion
software.

Stability: `0`
```

**Review by adamdmharvey — COMMENTED** [review]
Minor formatting; really like how this is working out and helps explain more about each plugin. Well done!

**Comment by Rugvip** [comment]
Removed a bunch of plugins from the list and just kept the core ones and some additional ones owned by Spotify teams. Encouraged PRs for plugin owners that want to add them to the list tho

File: docs/overview/stability-index.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+- **0** - There's a changelog entry.
+- **1** - There's a migration guide.
+- **2** - 2 weeks of deprecation.
+- **3** - 3 months of deprecation.
+
+## Packages
+
+### [`example-app`](https://github.com/backstage/backstage/tree/master/packages/app/)
+
+This is the `packages/app` package, and it's serves as an example as well as
```
Comment (adamdmharvey) [inline docs/overview/stability-index.md:44]:
```suggestion
This is the `packages/app` package, and it serves as an example as well as
```

oops typo

File: docs/overview/stability-index.md [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+utility for local development in the main Backstage repo.
+
+Stability: `N/A`
+
+### [`example-backend`](https://github.com/backstage/backstage/tree/master/packages/backend/)
+
+This is the `packages/backend` package, and it's serves as an example as well as
```
Comment (adamdmharvey) [inline docs/overview/stability-index.md:51]:
```suggestion
This is the `packages/backend` package, and it serves as an example as well as
```

**Comment by benjdlambert** [comment]
Can we merge this now we've shipped the latest version?

**Review by freben — APPROVED** [review]
